core/unit_test/incremental/Test02c_MemorySpace.hpp

```diff
@@ -0,0 +1,140 @@
+// the array copied back from the Device.
+
+namespace Test {
+
+using value_type       = double;
+const int num_elements = 10;
+const value_type value = 0.5;
+
+template <class ExecSpace>
+struct TestIncrMemorySpace_deepcopy {
+  // Memory space type for Device and Host data
+  using d_memspace_type = typename ExecSpace::memory_space;
+  using h_memspace_type = Kokkos::HostSpace;
+
+  // Routine to check correctness.
+  // Checks whether the original host array and the array copied/received from
+  // the Device are exactly the same.
+  void correctness_check(value_type *hostData, value_type *hostData_recv) {
+    // Check if the size of the originally allocated buffer and the buffer
+    // received from the Device are of equal length.
+    EXPECT_EQ(sizeof(hostData), sizeof(hostData_recv))
+        << "Original hostData and the copied back hostData_recv are of unequal "
+           "length";
+
+    // Check the element-wise equality of both the buffers.
+    for (int i = 0; i < num_elements; ++i) {
+      ASSERT_EQ(hostData[i], hostData_recv[i])
+          << "hostData and hostData_recv differ at index" << i;
+    }
+  }
+
+  // Routine to allocate memory in a specific memory space.
+  template <class MemSpace>
+  value_type *allocate_mem(int N) {
+    return (static_cast<value_type *>(
+        Kokkos::kokkos_malloc<MemSpace>("deviceData", N * sizeof(value_type))));
+  }
+
+  // Routine to free the memory from a specific memory space.
+  template <class MemSpace>
+  void free_mem(value_type *data) {
+    Kokkos::kokkos_free<MemSpace>(data);
+  }
+
+  void test_deepCopy() {
+    // Allocate memory on Device space.
+    auto *deviceData = allocate_mem<d_memspace_type>(num_elements);
+    ASSERT_NE(deviceData, nullptr);
+
+    // Allocate memory on Host space.
+    auto *hostData = allocate_mem<h_memspace_type>(num_elements);
+    ASSERT_NE(hostData, nullptr);
+
+    // Allocate memory on Host space which will receive data copied to the
+    // Device.
+    auto *hostData_recv = allocate_mem<h_memspace_type>(num_elements);
+    ASSERT_NE(hostData_recv, nullptr);
+
+    // Initialize Host data buffers with different values.
+    for (int i = 0; i < num_elements; ++i) {
+      hostData[i]      = (i + 1) * value;
+      hostData_recv[i] = num_elements - i;
+    }
+
+    // Copy first from Host to Device.
+    Kokkos::Impl::DeepCopy<h_memspace_type, d_memspace_type>(
+        deviceData, hostData, num_elements * sizeof(value_type));
+
+    // Copy from Device To the receive buffer of the Host.
+    Kokkos::Impl::DeepCopy<d_memspace_type, h_memspace_type>(
+        hostData_recv, deviceData, num_elements * sizeof(value_type));
+
+    // Check if all data has been copied correctly back to the Host.
+    correctness_check(hostData, hostData_recv);
+
+    // Free the allocated memory.
+    free_mem<d_memspace_type>(deviceData);
+    free_mem<h_memspace_type>(hostData);
+    free_mem<h_memspace_type>(hostData_recv);
```

Comment:
Is this better than calling `Kokkos::kokkos_free<MemSpace>(data);` directly?

Reply:
I had a different routine for free (even though it is just one line) here so that it is consistent with having a routine for kokkos_malloc too.